### PR TITLE
Fix strictness in `PlutusTx.IsData.TH`

### PR DESCRIPTION
**Pull request by zliu41**
Currently:

```haskell
let body =
        [|
            let !indexMatchCase = $(handleList argNames argsExpr)
                !fallthrough = $kont
            in BI.ifThenElse ($ixExpr `BI.equalsInteger` (index :: Integer)) (const indexMatchCase) (const fallthrough) BI.unitval
        |]
```

This is wrong - `indexMatchCase` and `fallthrough` should be non-strict, because (1) only one of them will be evaluated; (2) evaluating `indexMatchCase` when it shouldn't be evaluated can lead to `BI.head []` (e.g., in the `UnsafeFromData (Maybe a)` instance); (3) evaluating `fallthrough` when it shouldn't be evaluated can lead to PT1 (reconstructCaseError).

But why does it work now? Because bang patterns in TH quotations are ignored, so in the splice, there's no bang pattern on `indexMatchCase` or `fallthrough`.

For the other bang patterns, we do want them to be strict. This PR fixes them.

### CODE REVIEW

**Review by michaelpj — APPROVED** [review]
Good spot! needs a note explaining the issue, otherwise future people will never guess.

File: plutus-tx/src/PlutusTx/IsData/TH.hs [hunk 1]
```diff
@@ -44,16 +44,19 @@ reconstructCase (TH.ConstructorInfo{TH.constructorName=name, TH.constructorField
         handleList [] lExp = [| matchList $lExp $app (\_ _ -> Nothing) |]
         handleList (argName:rest) lExp = do
             tailName <- TH.newName "t"
+            consCaseName <- TH.newName "consCase"
             [|
-             let !consCase = \ $(TH.varP argName) $(TH.varP tailName) -> $(handleList rest (TH.varE tailName))
```
[the remark this answers is not in the document]
Comment (zliu41) [reply plutus-tx/src/PlutusTx/IsData/TH.hs:48]: I didn't find a ticket related to this. I'll make one.

**Comment by michaelpj** [comment]
What about the explanatory note?

**Comment by zliu41** [comment]
> What about the explanatory note?

Oh, right, I'll open a new PR for that.

**Pull request merged**